src/Compilers/CSharp/Portable/Binder/PatternExplainer.cs

```diff
@@ -0,0 +1,512 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Linq;
+using Microsoft.CodeAnalysis.CSharp.Symbols;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CSharp
+{
+    internal static class PatternExplainer
+    {
+        /// <summary>
+        /// Find the shortest path from the root node to the node of interest.
+        /// </summary>
+        /// <param name="nodes">The set of nodes in topological order.</param>
+        /// <param name="node">The node of interest.</param>
+        /// <param name="nullPaths">Whether to permit following paths that test for null.</param>
+        /// <returns>The shortest path, excluding the node of interest.</returns>
+        private static ImmutableArray<BoundDecisionDagNode> ShortestPathToNode(
+            ImmutableArray<BoundDecisionDagNode> nodes,
+            BoundDecisionDagNode node,
+            bool nullPaths)
+        {
+            // compute the distance from each node to the endpoint.
+            var dist = PooledDictionary<BoundDecisionDagNode, int>.GetInstance();
+            int nodeCount = nodes.Length;
+            int distance(BoundDecisionDagNode x) => x != null && dist.TryGetValue(x, out int v) ? v : nodeCount + 2;
```

Comment:
>dist.TryGetValue(x, out int v) [](start = 65, length = 30)

When does this fail? If it is expected to fail in some cases, can we assert something about those cases?

Reply:
It would fail for nodes not in `nodes`.

---
In reply to: [443066695](https://github.com/dotnet/roslyn/pull/44702#discussion_r443066695) [](ancestors = 443066695)